libmobilecoin/src/chacha20_rng.rs

```diff
@@ -0,0 +1,102 @@
+use crate::common::*;
+use mc_util_ffi::*;
+use rand_chacha::ChaCha20Rng;
+use rand_core::SeedableRng;
+use rand_core::RngCore;
+use std::sync::Mutex;
+use std::convert::TryInto;
+
+
+pub type McChaCha20Rng = ChaCha20Rng;
+
+pub struct McU128 {
+    pub bytes: [u8; 16]    
+}
+
+impl McU128 {
+    pub fn from_u128(val: u128) -> McU128 {
+        McU128 {
+            bytes: val.to_be_bytes(),
+        }
+    }
+
+    pub fn to_u128(&self) -> u128 {
+        u128::from_be_bytes(self.bytes)
+    }
+}
+
+impl IntoFfi<McU128> for McU128 {
+    #[inline]
+    fn error_value() -> McU128 {
+        McU128 {
+            bytes: [u8::MAX; 16],
+        }
+    }
+
+    #[inline]
+    fn into_ffi(self) -> McU128 {
+        self
+    }
+}
+
+impl_into_ffi!(FfiOwnedPtr<McU128>);
+impl_into_ffi!(Mutex<McChaCha20Rng>);
+
+#[no_mangle]
+pub extern "C" fn mc_chacha20_rng_create_with_long(long_val: u64) -> FfiOptOwnedPtr<Mutex<McChaCha20Rng>> {
+    ffi_boundary(|| {
+        Mutex::new(McChaCha20Rng::seed_from_u64(long_val))
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn mc_chacha20_rng_create_with_bytes(bytes: FfiRefPtr<McBuffer>) -> FfiOptOwnedPtr<Mutex<McChaCha20Rng>> {
+    ffi_boundary(|| {
+        let bytes: [u8; 32] = bytes.as_slice().try_into().expect("seed size must be 32 bytes");
+        Mutex::new(McChaCha20Rng::from_seed(bytes))
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn mc_chacha20_get_word_pos(
+    chacha20_rng: FfiMutPtr<Mutex<McChaCha20Rng>>,
+    out_word_pos: FfiMutPtr<McMutableBuffer>,
+) {
+    ffi_boundary(|| {
+        let word_pos = chacha20_rng.lock().unwrap().get_word_pos();
+        let mc_u128 = McU128::from_u128(word_pos);
+
+        let out_word_pos = out_word_pos
+            .into_mut()
+            .as_slice_mut_of_len(16)
+            .expect("word_pos length is not exaclty 16 bytes");
+
+        out_word_pos.copy_from_slice(&mc_u128.bytes);
+    })
```

Comment:
maybe we should just parameterize this call and not hold onto the chacha20 RNG reference in Swift land, we could just create a new one each time and get our result and then let go of it at the end of the function

Reply:
@the-real-adammork Just to make sure I understand... because currently the chacha20 instance maintains the state of the RNG for idempotence, if we don't retain it on the swift side in some manner, we must maintain the state of the seed + word pos for each RNG call in order to be able to recreate the prepared tx. How do you envision that side of things?

Reply:
Regarding the ordering of the RNG usage on the swift side, the use of the RNG is per transaction...meaning each transaction has its own rng, and the code that uses the rng has a fixed order of code execution - including the rust calls that bounce back to the swift side via the current implementation's use of the rng function that swift passes to iOS...so the swift side of things should be fine with respect to the usage of the rng.